### PR TITLE
Update new / removed entity GUI events

### PR DESCRIPTION
**Pull request by chapulina**
# 🦟 Bug fix

## Summary

Attempt of making the API a bit clearer after #1053 and #1138

* Made the distinction between GUI and server creation / removal 
* PIMPlized events in case we want to extend them later
* Made both events consistent, including both new and removed entities
* Updated names from "added" to "new" to be consistent with other APIs
* Added unit tests

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Updated documentation (as needed)
- [ ] ~~Updated migration guide (as needed)~~
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [x] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸🔸

### CODE REVIEW

File: src/gui/GuiEvents_TEST.cc [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+using namespace gui;
+
+/////////////////////////////////////////////////
+TEST(GuiEventsTest, GuiNewRemovedEntities)
+{
+  events::GuiNewRemovedEntities event({1, 2, 3}, {4, 5, 6});
+
+  EXPECT_LT(QEvent::User, event.type());
+
+  auto addedEntities = event.NewEntities();
```
Comment (adlarkin) [inline src/gui/GuiEvents_TEST.cc:34]:
```suggestion
  auto addedEntities = event.NewEntities();
  EXPECT_EQ(3u, addedEntities.size());
```
nit: we can add a check to verify that there are exactly 3 added entities

File: src/gui/GuiEvents_TEST.cc [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  EXPECT_NE(addedEntities.find(1), addedEntities.end());
+  EXPECT_NE(addedEntities.find(2), addedEntities.end());
+  EXPECT_NE(addedEntities.find(3), addedEntities.end());
+  EXPECT_EQ(addedEntities.find(100), addedEntities.end());
+
+  auto removedEntities = event.RemovedEntities();
```
Comment (adlarkin) [inline src/gui/GuiEvents_TEST.cc:40]:
```suggestion
  auto removedEntities = event.RemovedEntities();
  EXPECT_EQ(3u, removedEntities.size());
```

File: src/gui/GuiEvents_TEST.cc [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+/////////////////////////////////////////////////
+TEST(GuiEventsTest, NewRemovedEntities)
+{
+  events::NewRemovedEntities event({1, 2, 3}, {4, 5, 6});
+
+  EXPECT_LT(QEvent::User, event.type());
+
+  auto addedEntities = event.NewEntities();
```
Comment (adlarkin) [inline src/gui/GuiEvents_TEST.cc:54]:
```suggestion
  auto addedEntities = event.NewEntities();
  EXPECT_EQ(3u, addedEntities.size());
```

File: src/gui/GuiEvents_TEST.cc [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+  EXPECT_NE(addedEntities.find(1), addedEntities.end());
+  EXPECT_NE(addedEntities.find(2), addedEntities.end());
+  EXPECT_NE(addedEntities.find(3), addedEntities.end());
+  EXPECT_EQ(addedEntities.find(100), addedEntities.end());
+
+  auto removedEntities = event.RemovedEntities();
```
Comment (adlarkin) [inline src/gui/GuiEvents_TEST.cc:60]:
```suggestion
  auto removedEntities = event.RemovedEntities();
  EXPECT_EQ(3u, removedEntities.size());
```

**Review by adlarkin — APPROVED** [review]
Changes LGTM, thanks for cleaning this up a bit! I just left a few small comments about test coverage.

File: src/gui/GuiEvents_TEST.cc (hunk 1 above)
Comment (chapulina) [reply src/gui/GuiEvents_TEST.cc:34]: Good idea, see 43a014f765088c34157d9c4caecff6cb77fe4242

**Review by jennuine — APPROVED** [review]

File: include/ignition/gazebo/gui/GuiEvents.hh [hunk 5]
```diff
@@ -100,68 +102,60 @@ namespace events
...
-    /// \brief Set of recently removed entities
-    private: std::set<Entity> removedEntities;
+    /// \internal
+    /// \brief Private data pointer
+    IGN_UTILS_IMPL_PTR(dataPtr)
   };
 
-  /// \brief Event that notifies when new entities have been removed.
-  class RemovedEntities : public QEvent
+  /// \brief Event that notifies when new entities have been created and removed
```
Comment (nkoenig) [inline include/ignition/gazebo/gui/GuiEvents.hh:130]:
```suggestion
  /// \brief Event that notifies when new entities have been created or removed
```

File: include/ignition/gazebo/gui/GuiEvents.hh [hunk 6]
```diff
@@ -100,68 +102,60 @@ namespace events
     private: bool fromUser{false};
   };
 
-  /// \brief Event that contains newly created and removed entities
-  class AddedRemovedEntities : public QEvent
+  /// \brief Event that contains entities newly created and removed from the
```
Comment (nkoenig) [inline include/ignition/gazebo/gui/GuiEvents.hh:105]:
```suggestion
  /// \brief Event that contains entities newly created or removed from the
```

**Pull request merged**